incubator/hnc/pkg/reconcilers/hnc_config.go

```diff
@@ -38,33 +45,50 @@ type ConfigReconciler struct {
 // Reconcile sets up some basic variable and logs the Spec.
 // TODO: Updates the comment above when adding more logic to the Reconcile method.
 func (r *ConfigReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	// TODO: Surface the error more prominently and add a validating admission controller to prevent
-	// the problem in the first place.
+	ctx := context.Background()
+
+	// Validate the singleton name.
+	// TODO: Add a validating admission controller to prevent the problem in the first place.
 	if req.NamespacedName.Name != api.HNCConfigSingleton {
-		r.Log.Error(nil, "Singleton name is wrong. It should be 'config'.")
+		nnm := types.NamespacedName{Name: req.NamespacedName.Name}
+		inst := &api.HNCConfiguration{}
+		if err := r.Get(ctx, nnm, inst); err != nil {
+			r.Log.Error(err, "Couldn't read the singleton.")
+			return ctrl.Result{}, err
+		}
+
+		condition := api.HNCConfigurationCondition{
+			Code: api.CritSingletonNameInvalid,
+			Msg:  "Singleton name is wrong. It should be 'config'.",
+		}
+		inst.Status.Conditions = nil
+		inst.Status.Conditions = append(inst.Status.Conditions, condition)
+
+		if err := r.writeSingleton(ctx, inst); err != nil {
+			r.Log.Error(err, "Couldn't write singleton.")
+			return ctrl.Result{}, err
+		}
 		return ctrl.Result{}, nil
 	}
 
-	ctx := context.Background()
 	inst, err := r.getSingleton(ctx)
 	if err != nil {
 		r.Log.Error(err, "Couldn't read singleton.")
-		return ctrl.Result{}, nil
+		return ctrl.Result{}, err
 	}
 
 	// TODO: Modify this and other reconcilers (e.g., hierarchy and object reconcilers) to
 	// achieve the reconciliation.
 	r.Log.Info("Reconciling cluster-wide HNC configuration.")
 
-	// Create cooresponding ObjectReconcilers for newly added types, if needed.
-	// TODO: Returning an error here will make controller-runtime to requeue the object to be reconciled again.
-	// A better way to handle this is to display the error as part of the HNCConfig.status field to suggest the error is a
-	// permanent problem and to stop controller-runtime from retrying.
+	// Clear the existing conditions because we will reconstruct the latest conditions.
+	inst.Status.Conditions = nil
+
+	// Create corresponding ObjectReconcilers for newly added types, if needed.
 	// TODO: Rename the method syncObjectReconcilers because we might need more than creating ObjectReconcilers in future.
 	// For example, we might need to delete an ObjectReconciler if its corresponding type is deleted in the HNCConfiguration.
-	if err := r.createObjectReconcilers(inst); err != nil {
-		r.Log.Error(err, "Couldn't create the new object reconciler.")
-		return ctrl.Result{}, nil
+	if hasError := r.createObjectReconcilers(inst); hasError {
+		r.Log.Info("Object reconciler creation has an error.")
```

Comment:
Most of the error handling is already handled in `createObjectReconciler` so I'd log the problem there as well (you also have far more information there). Also, there's no need to name a temporary variable like this unless you want to refer to it again - though in this case, you can just change `createObjectReconcilers` so it doesn't return anything.

Reply:
Thanks for the suggestion. I moved the error log and condition change to `createObjectReconciler`. Changed both `createObjectReconcilers` and `creatObjectReconciler` to return nothing. Feel free to take a look.